juju/application.py

```diff
@@ -641,48 +643,85 @@ async def refresh(
         if switch is not None and revision is not None:
             raise ValueError("switch and revision are mutually exclusive")
 
-        charms_facade = client.CharmsFacade.from_connection(self.connection)
-        resources_facade = client.ResourcesFacade.from_connection(
-            self.connection)
+        resources_facade = client.ResourcesFacade.from_connection(self.connection)
+
         app_facade = self._facade()
 
-        charmstore = self.model.charmstore
-        charmstore_entity = None
+        charm_url = self.data['charm-url']
+        parsed_url = URL.parse(charm_url)
+        charm_name = parsed_url.name
+
+        # First we need to make sure we have the resources for the charm that's
+        # coming
+
+        # Get the list of resources needed to deploy this charm
+        if Schema.CHARM_HUB.matches(parsed_url.schema):
+            # Charmhub charms
+            charmhub = self.model.charmhub
+            charm_resources = await charmhub.list_resources(charm_name)
+
+            charm_url_origin_result = await app_facade.GetCharmURLOrigin(application=charm_name)
```

Comment:
GetCharmURLOrigin should be used for all charms, not just charmhub charms. 

An application name does not have to be the same as a charm name. Does pylibjuju not allow for an application name different from the charm name?